packages/examples/main/src/react/app.tsx

```diff
@@ -92,31 +81,41 @@ export const ReactMonacoEditor: React.FC<EditorProps> = ({
 
     useEffect(() => {
         if (ref.current != null) {
-            // register Monaco languages
-            monaco.languages.register({
-                id: 'json',
-                extensions: ['.json', '.jsonc'],
-                aliases: ['JSON', 'json'],
-                mimetypes: ['application/json']
-            });
-
-            // create Monaco editor
-            editorRef.current = monaco.editor.create(ref.current!, {
-                model: monaco.editor.createModel(defaultCode, 'json', monaco.Uri.parse('inmemory://model.json')),
-                glyphMargin: true,
-                lightbulb: {
-                    enabled: true
-                },
-                automaticLayout: true
-            });
-
-            // install Monaco language client services
-            MonacoServices.install();
-
-            lspWebSocket = createWebSocket(url);
+            const createEditor = () => {
+                // register Monaco languages
+                monaco.languages.register({
+                    id: 'json',
+                    extensions: ['.json', '.jsonc'],
+                    aliases: ['JSON', 'json'],
+                    mimetypes: ['application/json']
+                });
+
+                // create Monaco editor
+                editorRef.current = monaco.editor.create(ref.current!, {
+                    model: monaco.editor.createModel(defaultCode, 'json', monaco.Uri.parse('inmemory://model.json')),
+                    glyphMargin: true,
+                    lightbulb: {
+                        enabled: true
+                    },
+                    automaticLayout: true
+                });
+
+                lspWebSocket = createWebSocket(url);
+            };
+
+            const monacoVscodeApiInit = async () => {
+                await initServices({
+                    enableThemeService: true
+                }).then(() => createEditor());
+            };
+            if (init) {
+                monacoVscodeApiInit();
+            } else {
+                createEditor();
+            }
 
             return () => {
-                editorRef.current!.dispose();
+                editorRef.current?.dispose();
```

Comment:
nit: It's not a good practice to use a ref inside the effect dispose: better store the ref value at the effect start and use it here

Reply:
Thanks, I am not very deep into react.

Reply:
Implemented as suggested